group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java

```diff
@@ -0,0 +1,446 @@
+
+        processIdToState = new HashMap<>();
+        activeTaskToPrevOwner = new HashMap<>();
+        standbyTaskToPrevOwner = new HashMap<>();
+        for (Map.Entry<String, AssignmentMemberSpec> memberEntry : groupSpec.members().entrySet()) {
+            final String memberId = memberEntry.getKey();
+            final String processId = memberEntry.getValue().processId();
+            final Owner owner = new Owner(processId, memberId);
+            final AssignmentMemberSpec memberSpec = memberEntry.getValue();
+
+            processIdToState.putIfAbsent(processId, new ProcessState(processId));
+            processIdToState.get(processId).addMember(memberId);
+
+            // prev active tasks
+            for (Map.Entry<String, Set<Integer>> entry : memberSpec.activeTasks().entrySet()) {
+                Set<Integer> partitionNoSet = entry.getValue();
+                for (int partitionNo : partitionNoSet) {
+                    activeTaskToPrevOwner.put(new TaskId(entry.getKey(), partitionNo), owner);
+                }
+            }
+
+            // prev standby tasks
+            for (Map.Entry<String, Set<Integer>> entry : memberSpec.standbyTasks().entrySet()) {
+                Set<Integer> partitionNoSet = entry.getValue();
+                for (int partitionNo : partitionNoSet) {
+                    TaskId taskId = new TaskId(entry.getKey(), partitionNo);
+                    standbyTaskToPrevOwner.putIfAbsent(taskId, new HashSet<>());
+                    standbyTaskToPrevOwner.get(taskId).add(owner);
+                }
+            }
+        }
+    }
+
+    private GroupAssignment buildGroupAssignment(final Set<String> members) {
+        final Map<String, MemberAssignment> memberAssignments = new HashMap<>();
+
+        for (String memberId : members) {
+            Map<String, Set<Integer>> activeTasks = new HashMap<>();
+            if (activeTasksAssignments.containsKey(memberId)) {
+                activeTasks = buildTasks(activeTasksAssignments.get(memberId));
+            }
+            Map<String, Set<Integer>> standByTasks = new HashMap<>();
+            if (standbyTasksAssignments.containsKey(memberId)) {
+                standByTasks = buildTasks(standbyTasksAssignments.get(memberId));
+            }
+            memberAssignments.put(memberId, new MemberAssignment(activeTasks, standByTasks, new HashMap<>()));
+        }
+
+        return new GroupAssignment(memberAssignments);
+    }
+
+    private Map<String, Set<Integer>> buildTasks(final Set<TaskId> taskIds) {
+        Map<String, Set<Integer>> ret = new HashMap<>();
+        for (TaskId taskId : taskIds) {
+            ret.putIfAbsent(taskId.subtopologyId(), new HashSet<>());
+            ret.get(taskId.subtopologyId()).add(taskId.partition());
+        }
+        return ret;
+    }
+
+    private void assignActive(final Set<TaskId> activeTasks) {
+
+        // 1. re-assigning existing active tasks to clients that previously had the same active tasks
```

Comment:
Here you could have a queue for each of the steps. For example,
`activeTasksToReassignToPrevActiveOwner`, `activeTasksToReassignToPrevStandbyOwner`, and `activeTasksToAssignToAnyMember`. First all tasks are in `activeTasksToReassignToPrevActiveOwner`. All tasks are polled from the queue. Tasks that are not assigned in the first step are written to `activeTasksToReassignToPrevStandbyOwner`. All tasks are polled from the queue. Tasks that are not assigned in this step are written to `activeTasksToAssignToAnyMember`.
With a smart ordering of the tasks in the queues, we might be able to distribute the tasks well.

Reply:
>With a smart ordering of the tasks in the queues, we might be able to distribute the tasks well.

Could you please explain more? How? What is smart ordering and what do you mean by well distributing?